### PR TITLE
Rebuild CRLs on secondary performance clusters post migration and on new/updated issuers

### PR DESCRIPTION
**Pull request by stevendpclark**
- Hook into the backend invalidation function so that secondaries are notified of
   new/updated issuer or migrations occuring on the primary cluster. Upon notification
   schedule a CRL rebuild to take place upon the next process to read/update the CRL
   or within the periodic function if no request comes in.

### CODE REVIEW

File: builtin/logical/pki/crl_util.go
Comment (stevendpclark) [reply builtin/logical/pki/crl_util.go:68]: hmm good point, I'll investigate further to see who sees the invalidation calls across the nodes within a cluster.

Comment (stevendpclark) [reply builtin/logical/pki/crl_util.go:68]: So the only main node of a performance replica cluster gets its `invalidate` backend method called, which sets the flag to perform a deferred rebuild of the CRL. So we shouldn't need to worry about a thundering herd at least for this use case.

Comment (stevendpclark) [reply builtin/logical/pki/crl_util.go:68]: So the above comment is actually wrong, with additional testing I found a bug that led me to the wrong conclusion, invalidate does actually get called on all hosts.

**Comment by stevendpclark** [comment]
The last force push of the last commit, was just fixing up a formatting issue.

**Review by kitography — COMMENTED** [review]
This looks good to me. I wouldn't have caught what @cipherboy saw, but since that seems correct, I'll hold off approving.

File: builtin/logical/pki/backend.go
Comment (cipherboy) [inline builtin/logical/pki/backend.go:268]: I forgot to mention this on the last request, but one thing I noticed during our call today is that `resolveIssuerCRLPath` needs to return `legacyCRLPath, nil` when the cluster isn't migrated -- do you want to handle this as part of this PR or should I handle that in a separate PR once this one merges?

Comment (stevendpclark) [reply builtin/logical/pki/backend.go:268]: Just pushed a commit for this.

Comment (cipherboy) [reply builtin/logical/pki/backend.go:268]: thank you!

File: builtin/logical/pki/cert_util.go
Comment (sgmiller) [inline builtin/logical/pki/cert_util.go:152]: Is this enough information for an operator to know why this log line is emitted?

File: builtin/logical/pki/crl_util.go
Comment (sgmiller) [inline builtin/logical/pki/crl_util.go:64]: Can we call this requestRebuild*If*ActiveNode?  I initially thought this somehow sent a request to the active node.

**Review by sgmiller — CHANGES_REQUESTED** [review]
Just a couple small questions

File: builtin/logical/pki/crl_util.go
Comment (cipherboy) [reply builtin/logical/pki/crl_util.go:51]: `rebuildIfForced` is called from a number of contexts/places where, if we aren't _asked_ to rebuild the CRL, we shouldn't (e.g., the period function and/or CRL fetching has a hook to this call -- because the invalidate function doesn't have access to the relevant data for managed keys). 

We'd need to pair a `RWLock` with a `bool`, which is equivalent to an `atomic` int-ish, I think?

File: builtin/logical/pki/crl_util.go
Comment (stevendpclark) [reply builtin/logical/pki/crl_util.go:64]: Yup much better name.

File: builtin/logical/pki/crl_util.go
Comment (stevendpclark) [reply builtin/logical/pki/crl_util.go:51]: Correct, what @cipherboy said. We are kinda trying to recreate a sync.Once with resettable behaviour along with the ability. So both a RWLock would not be good enough as the `invalidate` call mainly doesn't have the context (missing storage and more importantly mountpoint) to rebuild the CRL. A sync.Once wouldn't be good enough as we need to redo this operation/reset it on subsequent `invalidate` calls in a safe manner.

File: builtin/logical/pki/cert_util.go
Comment (stevendpclark) [reply builtin/logical/pki/cert_util.go]: Changed to 

`b.Logger().Info("Using legacy CA bundle as PKI migration has not completed.")`

File: builtin/logical/pki/crl_util.go
Comment (cipherboy) [reply builtin/logical/pki/crl_util.go:51]: I think from the earlier call, a `sync.Once` can be a pointer to a `sync.Once` that allows it to be "resettable", though now you need an atomic "write pointer" I think?

Comment (stevendpclark) [reply builtin/logical/pki/crl_util.go:51]: Correct, but our use case isn't a direct match either for a sync.Once as the initial state, say for a single node, would be to never execute under a read api call occurred. That's another reason I did not use an actual sync.Once as the initialization code would have had to execute something arbitrary and then allow it to be reset.

Comment (sgmiller) [reply builtin/logical/pki/crl_util.go:51]: Good enough for me.

**Review by sgmiller — APPROVED** [review]

**Pull request merged**
